cuqi/geometry/_geometry.py

```diff
@@ -702,32 +702,47 @@ def _compute_eigpairs(self, xnod, C_nu, sigma, M, N_GL):
 
 class StepExpansion(Continuous1D):
     '''
-    class representation of the step random field with 3 intermidiate steps
+    Class representation of the step random field with n_steps intermediate steps.
+
+    Parameters:
+    -----------
+    grid: ndarray
+        Grid points for the step expansion to be evaluated at. There can be more grid points than number of steps.
+
+    n_steps: int
+        Number of equidistant steps.
+
+    kwargs: keyword arguments
+        keyword arguments are passed to the initializer of :class:`~cuqi.geometry.Continuous1D`
     '''
-    def __init__(self, grid, axis_labels=['x'], **kwargs):
+    def __init__(self, grid, n_steps=3, **kwargs):
 
-        super().__init__(grid, axis_labels,**kwargs)
+        super().__init__(grid, **kwargs)
+        if n_steps > np.size(self.grid):
+            raise ValueError("n_steps must be smaller than the number of grid points")
 
+        self._n_steps = n_steps
         L = self.grid[-1]
-        self._idx1 = np.where( (self.grid>=0*L)&(self.grid<=0.333*L) )
-        self._idx2 = np.where( (self.grid>0.333*L)&(self.grid<=0.666*L) )
-        self._idx3 = np.where( (self.grid>0.666*L)&(self.grid<=L) )
+
+        self._indices = []
+        for i in range(self._n_steps):
+            start = i*L/self._n_steps
+            end = (i+1)*L/self._n_steps
+            self._indices.append( np.where((self.grid>=start)&(self.grid<=end))[0] )
```

Comment:
I would think only one of the lower and upper thresholds should include equality to avoid the risk of having a grid point assigned to mulitple intervals?

Reply:
Good point! I update that.